src/extension-tree-shaking.js

```diff
@@ -97,16 +100,127 @@ addStealExtension(function (loader) {
-				.then(function(){
-					return name;
-				});
+			if (load && load.metadata.treeShakable) {
+				return reexecuteIfNecessary
+					.call(loader, load, parentName)
+					.then(function() {
+						return name;
+					});
 			}
 			return name;
 		});
+	};
+
+	function getImportSpecifierPositionsPlugin(load) {
+		load.metadata.importSpecifiers = Object.create(null);
+		load.metadata.importNames = Object.create(null);
+
+		return {
+			visitor: {
+				ImportDeclaration: function(path, state) {
+					var node = path.node;
+					var specifier = node.source.value;
+					var loc = node.source.loc;
+					load.metadata.importSpecifiers[specifier] = loc;
+					load.metadata.importNames[specifier] = (
+						node.specifiers || []
+					).map(function(spec) {
+						return spec.imported && spec.imported.name;
+					});
+				}
+			}
+		};
+	}
+
+	function treeShakePlugin(loader, load) {
+		var notShakable = {
+			exit: function(path, state) {
+				state.treeShakable = false;
+			}
+		};
+
+		var notShakeableVisitors = {
+			ImportDeclaration: notShakable,
+			FunctionDeclaration: notShakable,
+			VariableDeclaration: notShakable
+		};
+
+		return {
+			visitor: {
+				Program: {
+					enter: function(path) {
+						var state = {};
+						path.traverse(notShakeableVisitors, state);
+						load.metadata.treeShakable =
+							state.treeShakable !== false;
+					}
+				},
+
+				ExportNamedDeclaration: function(path, state) {
+					if (load.metadata.treeShakable) {
+						var usedResult = determineUsedExports.call(
+							loader,
+							load
+						);
+
+						var usedExports = usedResult.used;
+						var allUsed = usedResult.all;
+
+						if (!allUsed) {
+							path.get("specifiers").forEach(function(path) {
+								var name = path.get("exported.name").node;
+								if (
+									!usedExports.has(name) &&
+									name !== "__esModule"
+								) {
+									path.remove();
+								}
+							});
+
+							if (path.get("specifiers").length === 0) {
+								path.remove();
+							}
+						}
+					}
+				}
+			}
+		};
 	}
 
-	// determineUsedExports is used with a Babel tree-shaking plugin.
-	loader.determineUsedExports = determineUsedExports;
+	function applyBabelPlugin(load) {
+		var pluginLoader = loader.pluginLoader || loader;
+
+		return pluginLoader.import("babel").then(function(mod) {
+			var transpiler = mod.__useDefault ? mod.default : mod;
+			var babel = transpiler.Babel || transpiler.babel || transpiler;
+
+			try {
+				return babel.transform(load.source, {
+					plugins: [
+						getImportSpecifierPositionsPlugin.bind(null, load),
+						treeShakePlugin.bind(null, loader, load)
+					]
+				}).code;
+			} catch (e) {
+				return Promise.reject(e);
+			}
+		});
+	}
+
+	var translate = loader.translate;
+	var es6RegEx = /(^\s*|[}\);\n]\s*)(import\s+(['"]|(\*\s+as\s+)?[^"'\(\)\n;]+\s+from\s+['"]|\{)|export\s+\*\s+from\s+["']|export\s+(\{|default|function|class|var|const|let|async\s+function))/;
```

Comment:
I have to steal this from the es6 extension, `load.metadata.format` is not available at this point and I wasn't sure if there was a reliable way to do detection here other than using the regex.

Reply:
Ah, yeah, maybe we can move this somewhere shared...